**Name:** Sydney Bandi

**Email Id:** bandisydney@gmail.com

### Set up

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
root = "C:/Users/archd/Downloads/"

df = pd.read_csv(root + "Adops & Data Scientist Sample Data - Q1 Analytics.csv")
df.head()

ts user_id country_id site_id
0  2019-02-01 00:01:24  LC36FC        TL6   N0OTG
1  2019-02-01 00:10:19  LC39B6        TL6   N0OTG
2  2019-02-01 00:21:50  LC3500        TL6   N0OTG
3  2019-02-01 00:22:50  LC374F        TL6   N0OTG
4  2019-02-01 00:23:44  LCC1C3        TL6   QGO3G

In [3]:
df.shape

(3553, 4)

### Question 1

Consider only the rows with country_id = "BDV" (there are 844 such rows). For each site_id, we can compute the number of unique user_id's found in these 844 rows. Which site_id has the largest number of unique users? And what's the number?

In [4]:
df1 = df[df.country_id == "BDV"]
df1.shape

(844, 4)

Finding total unique users

In [5]:
df1.groupby('site_id')['user_id'].nunique().reset_index().sort_values('user_id', ascending = False).\
    rename(columns = {'user_id':'total_unique_users'})

site_id  total_unique_users
1   5NPAU                 544
2   N0OTG                  90
0   3POLC                   2

**Answer:** Site id "5NPAU" has the largest number of unique users and the number is 544.

### Question 2

Between 2019-02-03 00:00:00 and 2019-02-04 23:59:59, there are four users who visited a certain site more than 10 times. Find these four users & which sites they (each) visited more than 10 times. (Simply provides four triples in the form (user_id, site_id, number of visits) in the box below.)

Converting data type to datetime

In [6]:
df.ts.dtypes

dtype('O')

In [7]:
df.ts = pd.to_datetime(df.ts)
df.head()

ts user_id country_id site_id
0 2019-02-01 00:01:24  LC36FC        TL6   N0OTG
1 2019-02-01 00:10:19  LC39B6        TL6   N0OTG
2 2019-02-01 00:21:50  LC3500        TL6   N0OTG
3 2019-02-01 00:22:50  LC374F        TL6   N0OTG
4 2019-02-01 00:23:44  LCC1C3        TL6   QGO3G

Getting necessary data

In [8]:
df2 = df[(df.ts > pd.to_datetime("2019-02-03 00:00:00")) & (df.ts < pd.to_datetime("2019-02-04 23:59:59"))]
df2.head()

ts user_id country_id site_id
1049 2019-02-03 00:02:31  LC3C7E        TL6   3POLC
1050 2019-02-03 00:03:09  LC3C7E        TL6   3POLC
1051 2019-02-03 00:03:46  LC3C7E        TL6   3POLC
1052 2019-02-03 00:04:12  LC3C7E        TL6   3POLC
1053 2019-02-03 00:04:25  LC3C7E        TL6   3POLC

In [9]:
df2.tail()

ts user_id country_id site_id
2075 2019-02-04 23:54:56  LC34B0        XA7   N0OTG
2076 2019-02-04 23:55:46  LC3DEA        TL6   N0OTG
2077 2019-02-04 23:56:12  LC06C3        TL6   N0OTG
2078 2019-02-04 23:56:54  LC06C3        TL6   N0OTG
2079 2019-02-04 23:57:03  LC06C3        TL6   N0OTG

In [10]:
df2.shape

(1031, 4)

Finding total visits

In [11]:
df2 = df2.groupby(['user_id', 'site_id'])['ts'].count().reset_index()
df2 = df2.rename(columns = {'ts':'total_visits'})
df2 = df2[df2.total_visits > 10]
df2.sort_values('total_visits', ascending = False)

user_id site_id  total_visits
417  LC3A59   N0OTG            26
3    LC06C3   N0OTG            25
493  LC3C9D   N0OTG            17
485  LC3C7E   3POLC            15

**Answer:** (LC3A59, N0OTG, 26), (LC06C3, N0OTG, 25), (LC3C9D, N0OTG, 17), (LC3C7E, N0OTG, 15)  

### Question 3

For each site, compute the unique number of users whose last visit (found in the original data set) was to that site. For instance, user "LC3561"'s last visit is to "N0OTG" based on timestamp data. Based on this measure, what are top three sites? (hint: site "3POLC" is ranked at 5th with 28 users whose last visit in the data set was to 3POLC; simply provide three pairs in the form (site_id, number of users).)

Getting last visits of users

In [12]:
df3 = df.loc[df.groupby('user_id')['ts'].idxmax()]
df3.head()

ts user_id country_id site_id
1422 2019-02-03 18:52:50  LC00C3        QLT   5NPAU
1767 2019-02-04 11:35:10  LC01C3        QLT   5NPAU
733  2019-02-02 14:14:44  LC05C3        BDV   5NPAU
3078 2019-02-07 01:16:12  LC06C3        TL6   N0OTG
2402 2019-02-05 19:06:42  LC07C3        BDV   5NPAU

Finding top last visited sites

In [13]:
df3.groupby('site_id')['user_id'].count().reset_index().sort_values('user_id', ascending = False).\
    rename(columns ={'user_id':'users'}).reset_index(drop = True)

site_id  users
0   5NPAU    992
1   N0OTG    561
2   QGO3G    289
3   GVOFK     42
4   3POLC     28
5   RT9Z6      2
6   EUZ/Q      1
7   JSUUP      1

**Answer:** (5NPAU, 992), (N0OTG, 561), (QGO3G, 289)

### Question 4

For each user, determine the first site he/she visited and the last site he/she visited based on the timestamp data. Compute the number of users whose first/last visits are to the same website. What is the number?

Getting user's first visit and last visit

In [14]:
df_last = df.loc[df.groupby('user_id')['ts'].idxmax()]
df_first = df.loc[df.groupby('user_id')['ts'].idxmin()]

In [15]:
df_last.head()

ts user_id country_id site_id
1422 2019-02-03 18:52:50  LC00C3        QLT   5NPAU
1767 2019-02-04 11:35:10  LC01C3        QLT   5NPAU
733  2019-02-02 14:14:44  LC05C3        BDV   5NPAU
3078 2019-02-07 01:16:12  LC06C3        TL6   N0OTG
2402 2019-02-05 19:06:42  LC07C3        BDV   5NPAU

In [16]:
df_first.head()

ts user_id country_id site_id
1422 2019-02-03 18:52:50  LC00C3        QLT   5NPAU
1767 2019-02-04 11:35:10  LC01C3        QLT   5NPAU
733  2019-02-02 14:14:44  LC05C3        BDV   5NPAU
526  2019-02-01 22:49:39  LC06C3        TL6   N0OTG
2402 2019-02-05 19:06:42  LC07C3        BDV   5NPAU

Merging first and last visit

In [17]:
df4 = pd.merge(df_first, df_last, on="user_id", suffixes=('_first', '_last'))
df4.head()

ts_first user_id country_id_first site_id_first  \
0 2019-02-03 18:52:50  LC00C3              QLT         5NPAU   
1 2019-02-04 11:35:10  LC01C3              QLT         5NPAU   
2 2019-02-02 14:14:44  LC05C3              BDV         5NPAU   
3 2019-02-01 22:49:39  LC06C3              TL6         N0OTG   
4 2019-02-05 19:06:42  LC07C3              BDV         5NPAU   

              ts_last country_id_last site_id_last  
0 2019-02-03 18:52:50             QLT        5NPAU  
1 2019-02-04 11:35:10             QLT        5NPAU  
2 2019-02-02 14:14:44             BDV        5NPAU  
3 2019-02-07 01:16:12             TL6        N0OTG  
4 2019-02-05 19:06:42             BDV        5NPAU

Finding out users who have same first and last visit

In [18]:
df4[df4.site_id_first == df4.site_id_last].shape

(1670, 7)

**Answer:** 1670 users have first and last visit to same website.